[FX] Fix python code generation for wrapped getattr() with default value

yqhu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64271

Summary:
Closes #60417

Modified emit_node() in fx/graph.py to generate getattr() call with default value when len(node.args) != 2 instead of accessing the attribute.
Added test_torch_fx_getattr() in test/test_fx.py.

Test Plan:
pytest

Reviewers:

Subscribers:

Tasks:
T97189071

Tags:
bootcamp

Differential Revision: [D30671265](https://our.internmc.facebook.com/intern/diff/D30671265)

yqhu committed:
[FX] Fix python code generation for wrapped getattr() with default value

Summary:
Closes #60417

Modified emit_node() in fx/graph.py to generate getattr() call with default value when len(node.args) != 2 instead of accessing the attribute.
Added test_torch_fx_getattr() in test/test_fx.py.

Test Plan:
pytest

Reviewers:

Subscribers:

Tasks:
T97189071

Tags:
bootcamp

[ghstack-poisoned]

yqhu commented:
@yqhu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30671265).

jamesr66a commented on test/test_fx.py:
```diff
@@ -95,6 +95,8 @@ def a_lifted_leaf2(a, b):
 
 wrap('len')
 
+wrap('getattr')
```
What if we do this by default inside tracing, e.g. like here: 
https://github.com/pytorch/pytorch/blob/0ef8760bf6b3e8098ef42df60f1e451234151f32/torch/fx/_symbolic_trace.py#L610

I'm primarily curious how this would affect the current semantics of tracing. If it breaks stuff we can skip it

yqhu replied:
I agree.  `getattr` is used extensively and I wasn't quite sure about any possible side effect of patching it.  I saw the builtin function `len` was wrapped so figured we could possibly do the same for `getattr`.

jamesr66a replied:
Can you run an experiment where you try to patch it here and run all of `test_fx.py` and see if it breaks anything?

yqhu replied:
Will do

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -926,7 +926,8 @@ def emit_node(node : Node):
+                   node.args[1].isidentifier() and \
+                   len(node.args) == 2:
                     # pretty print attribute access
                     body.append(f'{repr(node)}{maybe_type_annotation} = {_format_target(repr(node.args[0]), node.args[1])}')
```
So I think this is baking in the existence of the attribute at trace time rather than resolving it at runtime. Can you print out the `code` in the test case to confirm?

What I was thinking of was actually checking in the branches here if `len(args) == 2` and emitting an explicit `getattr` node in the generated code. Then, the existence of the attribute would be checked at runtime

yqhu replied:
This is the generated `code` in the test case:
`
def forward(self, x):
    _tensor_constant0 = self._tensor_constant0
    getattr_1 = getattr(x, 'nonexistent_attr', _tensor_constant0);  x = _tensor_constant0 = None
    return getattr_1
`
I think the first line after the `if` block,
`body.append(f'{repr(node)}{maybe_type_annotation} = {global_name}({_format_args(node.args, node.kwargs)})')`
generated the `getattr` call.  It's a good idea to check the existence of the attribute at runtime; I'll make that change.  Thanks.

yqhu replied:
Sorry the previous formatting was bad.  This is the output of `print(traced.code)`:

`def forward(self, x):`
`    _tensor_constant0 = self._tensor_constant0`
`    getattr_1 = getattr(x, 'nonexistent_attr', _tensor_constant0);  x = _tensor_constant0 = None`
`    return getattr_1`

jamesr66a reviewed (commented):
Cool! Couple comments

yqhu committed:
Update on "[FX] Fix python code generation for wrapped getattr() with default value"


Summary:
Closes #60417

Modified emit_node() in fx/graph.py to generate getattr() call with default value when len(node.args) != 2 instead of accessing the attribute.
Added test_torch_fx_getattr() in test/test_fx.py.

Test Plan:
pytest

Reviewers:

Subscribers:

Tasks:
T97189071

Tags:
bootcamp

Differential Revision: [D30671265](https://our.internmc.facebook.com/intern/diff/D30671265)

[ghstack-poisoned]

yqhu commented:
@yqhu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30671265).

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -923,11 +923,13 @@ def emit_node(node : Node):
                 qualified_name = _get_qualified_name(node.target)
                 global_name = add_global(qualified_name, node.target)
+                # special case for getattr: node.args could be 2-argument or 3-argument
+                # 2-argument: attribute access; 3-argument: fall through to attrib function call with default value
```
nit: To be more precise, this branch handles the case where we have two arguments _and_ the identifier name argument is a valid identifier (i.e. not something like `"0"` that comes from `nn.Sequential`.

yqhu replied:
This is good to know.  The fall through behavior seems to affect the readability of the code.

jamesr66a reviewed (approved):
Looks good! Go ahead and land. The other comments are nits or questions about experiments
